azkaban-common/src/main/java/azkaban/trigger/Condition.java

```diff
@@ -62,25 +73,24 @@ public static Condition fromJson(final Object obj) throws Exception {
 
     final Map<String, Object> jsonObj = (HashMap<String, Object>) obj;
     Condition cond = null;
-
+    List<Long> missedCheckTimes = new ArrayList<>();
     try {
-      final Map<String, ConditionChecker> checkers =
-          new HashMap<>();
+      final Map<String, ConditionChecker> checkers = new HashMap<>();
       final List<Object> checkersJson = (List<Object>) jsonObj.get("checkers");
       for (final Object oneCheckerJson : checkersJson) {
-        final Map<String, Object> oneChecker =
-            (HashMap<String, Object>) oneCheckerJson;
+        final Map<String, Object> oneChecker = (HashMap<String, Object>) oneCheckerJson;
         final String type = (String) oneChecker.get("type");
-        final ConditionChecker ck =
-            checkerLoader.createCheckerFromJson(type,
-                oneChecker.get("checkerJson"));
+        final ConditionChecker ck = checkerLoader.createCheckerFromJson(type, oneChecker.get("checkerJson"));
         checkers.put(ck.getId(), ck);
+        // there is only one BasicTimeChecker associated with each condition
+        if (ck instanceof BasicTimeChecker && ck.getId().equals("BasicTimeChecker_1")) {
+          missedCheckTimes = (((BasicTimeChecker) ck).getMissedCheckTimesBeforeNow());
+        }
```

Comment:
This implementation looks fragile as it relies on the name of an object (`BasicTimeChecker_1`). I wonder if we should add a "getMissedChecks" method to the ConditionChecker interface instead of to the specific checker BasicTimeChecker.

Reply:
(can be done in a separate PR)

Reply:
I agree this code is kind of ugly, but it's not fragile:
Trigger instance use only two condition. TriggerCondition and ExpireCondition. Both Condition Object we create "BasicTimeChecker" object to it, and they only one checker inside. 
Second, let's look at other ConditionCheckers: ExecutionChecker does not even have nextCheckTime Concept, and missedSchedule does not apply to SlaCheckers too. That's why I think missedChecks is only specific to BasicTimeChecker only.

Reply:
It looks like "getMissedChecks" method doesn't make sense for ALL the existing ConditionCheckers, understood. Thank you for considering adding the behavior to the interface!